### ISSUE TITLE
Improvement for upstream ssl_type

### ISSUE THREAD

**Issue by sadiqkhoja** [getodk/central#601]
#### Problem description

Currently user needs to modify `docker-compose.yml` if they want to use upstream SSL. I think we should pass `UPSTREAM_HTTPS_PORT` and `SSL_TYPE` environment variables to the [start-odk.sh](https://github.com/getodk/central/blob/master/files/service/scripts/start-odk.sh) and [start-enketo.sh](https://github.com/getodk/central/blob/master/files/enketo/start-enketo.sh). So that `BASE_URL` can be decided accordingly.

**Comment by alxndrsn** [comment]
Maybe something has changed, but I think `SSL_TYPE` **can** be passed as an environment variable.

1. `start-enketo.sh`: this file currently has no mention of `SSL_TYPE`
2. `start-odk.sh`: this file is run at container startup, so changing `SSL_TYPE` will be effective on container restart

I don't think `SSL_TYPE` has any effect at container build time.

This doesn't seem to be a significant change from 106f533a128d43bb71ff2366bf2d07ff23082b38, which would have been the head of the `next` branch at the time this issue was originally filed.

@sadiqkhoja can you clarify?

**Comment by sadiqkhoja** [comment]
Please see https://docs.getodk.org/central-install-digital-ocean/#using-upstream-ssl 

If we pass `UPSTREAM_HTTPS_PORT` and `SSL_TYPE` environment variables to the `service` and `enketo` services/containers in docker-compose.yml then we add a condition like: if SSL_TYPE is `upstream` then use UPSTREAM_HTTPS_PORT otherwise `HTTPS_PORT` in the following line in both startup files:

`BASE_URL=$( [ "${HTTPS_PORT}" = 443 ] && echo https://"${DOMAIN}" || echo https://"${DOMAIN}":"${HTTPS_PORT}" ) `

In this issue (a wish) my intention is to eliminate the need of modifying `docker-compose.yml` so that administrators could run Central behind an upstream proxy by just setting their config in `.env` file. Hence avoiding any potential git conflict.

**Comment by alxndrsn** [comment]
@sadiqkhoja thanks for the update.

I think there are two separate types of URLs which would ideally be distinguished:

1. internal communications between `service` and `enketo`.  These should use internal docker service names, and port numbers (whether HTTP or HTTPS) can be hard-coded
2. external-facing URLs for client use.  These should be able to use the current `${DOMAIN}` or `${DOMAIN}:${HTTPS_PORT}` approach.

I don't think this approach would require any changes to configuration.

Would that work?

**Comment by sadiqkhoja** [comment]
I agree that there has to be two separate types of URLs. For the second case, I think we still need `UPSTREAM_HTTPS_PORT` because it can be different that 443.

**Comment by alxndrsn** [comment]
> I think we still need `UPSTREAM_HTTPS_PORT` because it can be different that 443.

When would `UPSTREAM_HTTPS_PORT` use differ to `HTTPS_PORT`?

**Comment by sadiqkhoja** [comment]
- `UPSTREAM_HTTPS_PORT` would be the port on which upstream proxy will be listening
- `HTTPS_PORT` will be Central's nginx port - but in case of upstream ssl, I think Central's nginx should be listening on only non-ssl port (http_port).

if you think we can get away with only one variable i.e. `HTTPS_PORT` then that would be great.

**Comment by alxndrsn** [comment]
> * `UPSTREAM_HTTPS_PORT` would be the port on which upstream proxy will be listening
> 
>     * `HTTPS_PORT` will be Central's nginx port - but in case of upstream ssl, I think Central's nginx should be listening on only non-ssl port (http_port).
> 
> 
> if you think we can get away with only one variable i.e. `HTTPS_PORT` then that would be great.

If docker services communicate via docker network, then `HTTPS_PORT` only needs to refer to the publicly-facing HTTPS port.

**Comment by sadiqkhoja** [comment]
got it.
